Sources/ProcedureKit/ProcedureQueue.swift

```diff
@@ -137,129 +184,273 @@ open class ProcedureQueue: OperationQueue {
+                promise.complete()
+            }
 
             return
         }
 
+        // Procedure subclass
+
         procedure.log.verbose(message: "Adding to queue")
 
         /// Add an observer to invoke the will finish delegate method
-        procedure.addWillFinishBlockObserver { [weak self] procedure, errors in
+        procedure.addWillFinishBlockObserver { [weak self] procedure, errors, futureFinish in
             if let queue = self {
-                queue.delegate?.procedureQueue(queue, willFinishOperation: procedure, withErrors: errors)
+                queue.delegate?.procedureQueue(queue, willFinishProcedure: procedure, withErrors: errors)?.then(on: queue.dispatchQueue) {
+                    // ensure that the observed procedure does not finish prior to the
+                    // willFinishProcedure delegate completing
+                    futureFinish.doThisBeforeEvent()
+                }
             }
         }
 
         /// Add an observer to invoke the did finish delegate method
         procedure.addDidFinishBlockObserver { [weak self] procedure, errors in
             if let queue = self {
-                queue.delegate?.procedureQueue(queue, didFinishOperation: procedure, withErrors: errors)
+                queue.delegate?.procedureQueue(queue, didFinishProcedure: procedure, withErrors: errors)
             }
         }
 
-        /// Process any conditions
-        if procedure.conditions.count > 0 {
+        // Indicate to the operation that it is to be enqueued
+        procedure.willEnqueue(on: self)
 
-            /// Check for mutual exclusion conditions
-            let manager = ExclusivityManager.sharedInstance
+        if let delegate = delegate {
 
-            let mutuallyExclusiveConditions = procedure.conditions.filter { $0.isMutuallyExclusive }
-            var previousMutuallyExclusiveOperations = Set<Operation>()
+            // WillAddProcedure delegate
+            (delegate.procedureQueue(self, willAddProcedure: procedure, context: context) ?? _SyncAlreadyAvailableFuture()).then(on: dispatchQueue) { _ in
 
-            for condition in mutuallyExclusiveConditions {
-                let category = "\(condition.category)"
-                if let previous = manager.add(procedure: procedure, category: category) {
-                    previousMutuallyExclusiveOperations.insert(previous)
-                }
+                // Step 2:
+                self._add_step2(procedure: procedure, context: context, promise: promise)
+            }
+        }
+        else {
+
+            // if no delegate, proceed to Step 2 directly:
+            _add_step2(procedure: procedure, context: context, promise: promise)
+        }
+    }
+
+    // Step2 2: process conditions
+    //          add the evaluator procedure to self (using async method), .then(onQueue: )
+    //
+    private func _add_step2(procedure: Procedure, context: Any?, promise: ProcedurePromise) {
```

Comment:
Yeah, I agree with @jshier here. I think things like "step 2" are ill-defined, and will be hard to read & maintain in 3 months time.

Not sure whether renaming is something we can do a bit later though, as I think we can tackle style/naming later on, after merging, to avoid as much risk.